### ISSUE TITLE
Kedro loss lightgbm model paramater due to deepcopy(lightgbm_model) 

### ISSUE THREAD

**Issue by noklam** [quantumblacklabs/kedro#730]
## Description
When loading lightGBM with Memory Dataset, the parameters are lost due to deepcopy.

This is not a bug from Kedro, but I think Kedro should consider changing the default to prevent this happens. The issue is related to `deepcopy(lgb_model)` causing parameters lsot.

## Context
How has this bug affected you? What were you trying to accomplish?

## Steps to Reproduce
git clone https://github.com/noklam/kedro_lightgbm_bug

kedro run
```
In the `create_model()` node
> {'objective': 'regression', 'verbose': -1, 'num_leaves': 3} # prints parameters

In the `load_model()` node
>{} # prints parameters, it lost all the parameters
```

python test_lightgbm.py (without involving kedro, you can see the different output from deepcopy(model) versus pickle.load)
```
import pandas as pd
import numpy as np
import pickle
import lightgbm as lgb
from copy import deepcopy

params = {
'objective': 'regression',
'verbose': -1,
'num_leaves': 3
}

X = np.random.rand(100,2)
Y = np.ravel(np.random.rand(100,1))
lgbm = lgb.train(params, lgb.Dataset(X,label=Y),num_boost_round=1)

f = open('test_pickle.pkl','wb')
pickle.dump(lgbm,f)
f.close()

print(lgbm.params)

## Deep copy will missing params
new_model = deepcopy(lgbm)
print(new_model.params)


## Load from file is fine
import pickle
f = open('test_pickle.pkl','rb')
m2 = pickle.load(f)
f.close()

print(lgbm.params)

```

> {'objective': 'regression', 'verbose': -1, 'num_leaves': 3}
> {}
> {'objective': 'regression', 'verbose': -1, 'num_leaves': 3}

## Expected Result
parameters should be preserved.

## Actual Result
paramteres lost, it becomes empty dict.

```
-- If you received an error, place it here.
```

```
-- Separate them if you have more than one.
```

## Your Environment
Include as many relevant details about the environment in which you experienced the bug:

* Kedro version used (`pip show kedro` or `kedro -V`): 0.16.6
* Python version used (`python -V`): 3.7.5
* Operating system and version: Windows 10

**Comment by noklam** [comment]
I have done some investigation related to this issue in Lightgbm repository.

https://github.com/microsoft/LightGBM/issues/4085

**Comment by AntonyMilneQB** [comment]
HI @noklam, thanks for all this, and especially for making the demo - it makes it really easy to try out 🙂  As you say, this seems to be a LightGBM problem rather than kedro, and so I am wondering what we might do about it. It seems that using this in the catalog works correctly to preserve the parameters, right?
```
TRAINED_MODEL:
    type: MemoryDataSet
    copy_mode: assign
```

What sort of fix did you have in mind for this on the kedro side? We could add some special logic in `MemoryDataSet` to not try and do `deepcopy` on a LightGBM model. But this doesn't seem good since we'd have to remove it whenever LightGBM have fixed it, and then whether or not to use the special logic would depend on the LightGBM version, etc. Currently `MemoryDataSet` is very light and doesn't contain special cases like this, so I suspect that we'd prefer to avoid adding one, especially if the temporary solution of using `copy_mode: assign` works ok.

**Comment by AntonyMilneQB** [comment]
P.S. might also be worth trying out [sklearn.base.clone](https://scikit-learn.org/stable/modules/generated/sklearn.base.clone.html) - not sure if that will work well with LightGBM or not

**Comment by AntonyMilneQB** [comment]
P.P.S. I also wonder if this behaviour is for some reason intentional in LightGBM. Like you say they have explicitly defined `__deepcopy__` to only copy `Booster`. Maybe there is a good reason for that? Will be interesting to see what they say.

**Comment by noklam** [comment]
For kedro user, they do not necessarily know that deepcopy was used under the hood, and thus this may not be the expected behavior.

I don't have a nice solution in mind, as you said it will be a special case checking for MemoryDataset. I think users needed to be well inform so they know that they should set the assign mode manually.


On the other hand, we need to wait for lightgbm side why they implement the deepcopy method in this way, so there is no guarantee they would fix this at the moment.


I am actually quite surprised this issue was not mentioned in both kedro & lightgbm repository

**Comment by AntonyMilneQB** [comment]
> For kedro user, they do not necessarily know that deepcopy was used under the hood, and thus this may not be the expected behavior.

This is very true, and it definitely seems like undesirable behaviour (unless I've misunderstood something about LightGBM models).

> I am actually quite surprised this issue was not mentioned in both kedro & lightgbm repository

Me too. I guess in kedro people tend to persist models so might not have come across this issue before. I am surprised if this problem has existed in LightGBM and no one has noticed it before though 🤔  On the whole, I think let's wait and see what LightGBM says on the matter - maybe there's actually a good reason they implemented `deepcopy` like this.

**Comment by AntonyMilneQB** [comment]
I see LightGBM just responded. As I understand it, actually `deepcopy` is working correctly - it doesn't pass the parameters to the Python wrapper correctly at the moment, but this doesn't affect the model. So actually having `deepcopy` being the default for a `MemoryDataSet` should still work fine in kedro.

**Comment by noklam** [comment]
Seems it is rather a workaround to make lightgbm pickable(I do not know much about this, parameter is just a dict, I may need to read more to understand the issue later).

It is more desirable to keep the parameters with the model, but I agree at this point there is nothing to do for Kedro.

Not sure if they are going to fix this issue in next release. It's not going to affect model inference, but parameters are still potentially useful for other downstream tasks.

Thanks for following the issue, I think it is ok to close the issue.

**Comment by noklam** [comment]
I wrote a casual summary here in case someone interest in this issue.
https://noklam.ml/python/pickle/deepcopy/2021/03/19/deepcopy,-lightgbm-and-Pickles.html#%F0%9F%A5%96Italian-BMT,-%F0%9F%A5%ACLettuce--%F0%9F%8D%85-tomato-and-some-%F0%9F%A5%92pickles-please
